errors/context.go

```diff
@@ -0,0 +1,193 @@
+package errors
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/mailgun/holster/v5/callstack"
+	"github.com/sirupsen/logrus"
+)
+
+// HasContext Implement this interface to pass along unstructured context to the logger
+type HasContext interface {
+	Context() map[string]interface{}
+}
+
+// HasFormat True if the interface has the format method (from fmt package)
+type HasFormat interface {
+	Format(st fmt.State, verb rune)
+}
+
+// WithContext Creates errors that conform to the `HasContext` interface
+type WithContext map[string]interface{}
+
+// Wrapf returns an error annotating err with a stack trace
+// at the point Wrapf is call, and the format specifier.
+// If err is nil, Wrapf returns nil.
+func (c WithContext) Wrapf(err error, format string, args ...interface{}) error {
+	if err == nil {
+		return nil
+	}
+	return &withContext{
+		stack:   callstack.New(1),
+		context: c,
+		cause:   err,
+		msg:     fmt.Sprintf(format, args...),
+	}
+}
+
+// Wrap returns an error annotating err with a stack trace
+// at the point Wrap is called, and the supplied message.
+// If err is nil, Wrap returns nil.
+func (c WithContext) Wrap(err error, msg string) error {
+	if err == nil {
+		return nil
+	}
+	return &withContext{
+		stack:   callstack.New(1),
+		context: c,
+		cause:   err,
+		msg:     msg,
+	}
+}
+
+func (c WithContext) Error(msg string) error {
+	return &withContext{
+		stack:   callstack.New(1),
+		context: c,
+		cause:   errors.New(msg),
+		msg:     "",
+	}
+}
+
+func (c WithContext) Errorf(format string, args ...interface{}) error {
+	return &withContext{
+		stack:   callstack.New(1),
+		context: c,
+		cause:   fmt.Errorf(format, args...),
+		msg:     "",
+	}
+}
+
+// Implements the `error` `causer` and `Contexter` interfaces
+type withContext struct {
+	context WithContext
+	msg     string
+	cause   error
+	stack   *callstack.CallStack
+}
+
+func (c *withContext) Unwrap() error {
+	return c.cause
+}
+
+func (c *withContext) Is(target error) bool {
+	_, ok := target.(*withContext)
+	return ok
+}
+
+func (c *withContext) Error() string {
+	if c.msg == "" {
+		return c.cause.Error()
+	}
+	return c.msg + ": " + c.cause.Error()
+}
+
+func (c *withContext) StackTrace() callstack.StackTrace {
```

Comment:
This makes every error in a chain return the stacktrace of the root cause error, but that is not always what we want. But that makes us lose information about where each of the errors in chain was wrapped. This why when we log `WithError` we get several stacktraces at each point of wrap.

Reply:
@horkhe Just use wrapping.
The problem with stack trace is that if the code is changed often the file_name:line from stack trace might point to a totally different piece of code. With good error messages, it should be more stable.